src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java

```diff
@@ -73,9 +77,9 @@ public ElementNode(AbstractNode root, AbstractNode parent, DetailAST detailAst)
         this.parent = parent;
         this.root = root;
         this.detailAst = detailAst;
-        setIdent();
-        createChildren();
         text = TokenUtil.getTokenName(detailAst.getType());
+        createAttributes();
```

Comment:
Can we delay creating attributes until they are needed to prevent doing excess work?

Maybe speed should be looked at separately, but this seems like it would create similar problems we are already seeing for stackoverflow on children. See https://github.com/checkstyle/checkstyle/issues/5820 .
If attributes are only used when the user specifies them in their XPath query, then it makes more sense to only create them for the nodes users specify them for.

Reply:
Xpath AST is created for all for queries before matching certain query. We can not predict how queries will look like.

please clarify your concern.

Reply:
I debugged `getAttributeValue` some, and yes, it is only called if the xpath has an attribute. So we don't need to populate it for all nodes if we don't always use it.

It seems `iterateAxis` is used very rarely in our tests (mostly `XpathMapperTest`). I see it being invoked when `*` and possibly `//` is used, but it still shows it doesn't need the attributes for every node depending on the xpath given by the user.

Reply:
If you want to move it to a new issue that is fine.
`createAttributes` method is new and it is getting worse with the map of attributes but we had `setIdent` and `text` before.

Reply:
@timurt , please step in.

Reply:
@rnveach @romani 

> I debugged getAttributeValue some, and yes, it is only called if the xpath has an attribute. So we don't need to populate it for all nodes if we don't always use it.

Didn't really understand, do you suggest to use old `setIdent` method inside `getAttributeValue ` method and store result for next queries?

> It seems iterateAxis is used very rarely in our tests (mostly XpathMapperTest). I see it being invoked when * and possibly // is used, but it still shows it doesn't need the attributes for every node depending on the xpath given by the user.

About `iterateAxis`, we can create `AttributeNode` array inside method during runtime

@rnveach 
If I understood you correctly, you want lazy loading, data is created only when it is needed?

I can check these approaches with [5820](https://github.com/checkstyle/checkstyle/issues/5820) issue and see if it will help or not

Reply:
> you want lazy loading, data is created only when it is needed

Yes that is correct.

> I can check these approaches with 5820 

That is fine.

Reply:
@rnveach @romani 

> I can check these approaches with 5820

`StackOverflow` happens because AST tree of this [file](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/api/detailast/InputDetailASTJustToMakeStackoverflowError.java) has the following AST structure:

I couldn't even print full AST tree of the file, because it takes a very long time for CMD tool.
```
    |   |     `--PLUS -> + [3:96]
    |   |           |--PLUS -> + [3:93]
    |   |           |   |--PLUS -> + [3:90]
    |   |           |   |   |--PLUS -> + [3:87]
    |   |           |   |   |   |--PLUS -> + [3:84]
    |   |           |   |   |   |   |--PLUS -> + [3:81]
...
    |   |           |   |   |   |   |   `--STRING_LITERAL -> "" [3:82]
    |   |           |   |   |   |   `--STRING_LITERAL -> "" [3:85]
    |   |           |   |   |   `--STRING_LITERAL -> "" [3:88]
    |   |           |   |   `--STRING_LITERAL -> "" [3:91]
    |   |           |   `--STRING_LITERAL -> "" [3:94]
    |   |           `--STRING_LITERAL -> "" [3:97]
```
See, every `PLUS` node has two child nodes: `PLUS` and `STRING_LITERAL`, 
the problem we are making a lot of recursive calls (not infinite).

We cannot avoid creating children nodes, otherwise we won't be able to query them.
Making `getAttributeValue` lazy does not help - because `StackOverflow` error appears on `Xpath tree initialization` phase, not on query phase.

Currently I see only one way to solve this problem, switch from `DFS` to `BFS`.
Remove recursive calls and use queues:
1. Take node from queue
2. Process it
3. Put child nodes into the queue
4. Move queue cursor to the next element

Btw, if you want I can implement lazy approach, but I am not sure what it will give us

Reply:
> We cannot avoid creating children nodes

Lazy initialization isn't about not creating the nodes, it is about putting the creation at another time. We have stackoverflow because we are doing too may createChildren at once, all chained together. As long as we do 1 create and not chain others in it, then stackoverflow should be resolved. It is the difference between a for loop and a recursive call.

Regardless, this out of the scope of this PR and will be done in the other issue. Lets finish this PR.

Reply:
Let's finish this PR, lazy and stackoverflow can be addressed later on.